File: Learning/project/Go/homi-server/model/reward.go
package data

import (
	"math"
)

type RewardType int8

const (
	GoldReward RewardType = iota
	ExpReward
	RoomSkinReward
	RoomStarReward
	MotionReward
	ScoreReward
)

type Bonus struct {
	Rate float64 `json:"rate" bson:"rate"`
	Val  float64 `json:"val" bson:"val"`
}

type Reward struct {
	Type     RewardType  `json:"type" bson:"type"`
	Value    float64     `json:"value" bson:"value"`
	Bonus    Bonus       `json:"bonus" bson:"bonus"`
	Params   interface{} `json:"params" bson:"params"` ///暂时没有作用
	Receiver string      `json:"receiver" bson:"receiver"`
}

func NewReward(type1 RewardType, value float64, rec string) *Reward {
	return &Reward{Type: type1, Value: value, Receiver: rec}
}

func (r Reward) RealValue() float64 {
	rate := 0.0
	if r.Bonus.Rate != 0.0 {
		rate = r.Bonus.Rate
	}
	bVal := 0.0
	if r.Bonus.Val != 0.0 {
		bVal = r.Bonus.Val
	}
	return math.Round(r.Value*(rate+1) + bVal)
}

func (r *Reward) BonusUp(rate float64, val float64) {
	r.Bonus.Val += val
	r.Bonus.Rate += rate
}

//// region 奖励Group处理
//
//type RewardGroup struct {
//	Rewards []*Reward `json:"rewards" bson:"rewards"`
//}
//
//func CreateRewardGroup(rs ...*Reward) *RewardGroup {
//	if rs == nil || len(rs) == 0 {
//		return &RewardGroup{Rewards: make([]*Reward, 0)}
//	}
//	res := &RewardGroup{rs}
//	return res
//}
//
//func (r *RewardGroup) Add(rs ...*Reward) {
//	if rs == nil || len(rs) == 0 {
//		return
//	}
//	for _, v := range rs {
//		r.Rewards = append(r.Rewards, v)
//	}
//}
//
////endregion

//type RewardProcessor interface {
//	//PreProcess() error        //预处理
//	//Conditions() []*Condition //附带条件
//	//Check()                   //消耗
//
//	Invoke(rewardType RewardType) error            //进行处理
//}
//

File: develop/Framework/src/modules/reward/Reward.ts
import {BaseData, field} from "../../data/BaseData";
import {createProcessor} from "./RewardProcessor";

export enum RewardType {
	Gold = "gold",
	Exp = "exp",
	// TODO: 枚举奖励类型
}

export function r(type: RewardType, value: number, params = {}) {
	return Reward.create(type, value, params);
}

export class Reward extends BaseData {

	@field
	public type: RewardType = RewardType.Gold;
	@field
	public value: number = 0;
	@field(Object)
	public bonus: {rate: number, val: number} = {rate: 0, val: 0};
	@field(Object)
	public params: object = {};

	public get realValue() {
		const rate = this.bonus.rate || 0;
		const val = this.bonus.val || 0;
		return Math.round(this.value * (rate + 1) + val)
	}

	public static create(type: RewardType,
											 value: number, params = {}) {
		const res = new Reward();
		res.type = type;
		res.value = value;
		res.params = params;
		return res;
	}

}



export class RewardGroup extends BaseData {

	@field([Reward])
	public rewards: Reward[] = [];

	public static create(...rewards: (Reward | RewardGroup)[]) {
		const res = new RewardGroup();
		rewards.forEach(reward => res.add(reward));
		return res;
	}

	public add(reward: Reward | RewardGroup) {
		if (reward instanceof Reward) {
			const curCond = this.find(reward.type);
			if (curCond) curCond.value += reward.value;
			else this.rewards.push(reward.clone());
		} else
			reward.rewards.forEach(c => this.add(c))
	}

	public push(reward: Reward) {
		this.rewards.push(reward);
	}

	public async invoke(openid, rate = 1) {
		const processors = this.rewards.map(
			r => createProcessor(openid, r));

		for (const p of processors) await p.check();
		for (const p of processors) await p.invoke(rate);
	}

	public find(type: RewardType) {
		return this.rewards.find(c => c.type == type);
	}

	public bonus(type: RewardType, rate?, val?) {
		this.find(type).bonus.val += val;
		this.find(type).bonus.rate += rate;
	}

	// region 快捷查找

	public get gold() { return this.find(RewardType.Gold); }
	public get exp() { return this.find(RewardType.Exp); }

	// endregion
}
